File: scenes/binary_search_table.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<stdint.h>

typedef uint8_t VALUE;

typedef char KEY;

int compare(KEY k1, KEY k2) {
	return k1 - k2;
}

typedef struct {
	uint32_t capacity;
	KEY* keys;
	VALUE* values;
	uint8_t size;
} BinarySearchTable;

int initialize(BinarySearchTable* p_bst, uint32_t capacity) {
	p_bst->capacity = capacity;
	p_bst->keys = malloc(capacity * sizeof(KEY));
	p_bst->values = malloc(capacity * sizeof(VALUE));
	p_bst->size = 0;
	return 0;
}

int rank(BinarySearchTable* p_bst, KEY key) {
	int low = 0, high = p_bst->size - 1, mid = 0;
	if (p_bst->size == 0)
		return 0;
	while (low <= high) {
		mid = (low + high) / 2;
		if (compare(p_bst->keys[mid], key) == 0) {
			return mid;
		}
		if (compare(p_bst->keys[mid], key) > 0) {
			high = mid - 1;
		}
		else {
			low = mid + 1;
		}
	}
	return low;
}

int put(BinarySearchTable* p_bst, KEY key, VALUE val) {
	int key_rank = 0, ix = 0;
	if (p_bst->size == 0 || compare(key, p_bst->keys[p_bst->size - 1]) > 0) {
		p_bst->keys[p_bst->size] = key;
		p_bst->values[p_bst->size] = val;
		p_bst->size++;
		return 1;
	}
	key_rank = rank(p_bst, key);
	if (compare(p_bst->keys[key_rank], key) == 0) {
		p_bst->values[key_rank] = val;
		return p_bst->size;
	}
	else {
		if (p_bst->size == p_bst->capacity) {
			return -1;
		}
		for (ix = p_bst->size; ix > key_rank; ix--) {
			p_bst->keys[ix] = p_bst->keys[ix - 1];
			p_bst->values[ix] = p_bst->values[ix - 1];
		}
		p_bst->keys[key_rank] = key;
		p_bst->values[key_rank] = val;
		p_bst->size += 1;
		return p_bst->size;
	}
}

int delete(BinarySearchTable* p_bst, KEY key) {
	int key_rank = 0, ix = 0;
	if (p_bst->size == 0 || compare(key, p_bst->keys[p_bst->size - 1]) > 0) {
		return 0;
	}
	key_rank = rank(p_bst, key);
	if (compare(p_bst->keys[key_rank], key) == 0) {
		for (ix = key_rank; ix < p_bst->size - 1; ix++) {
			p_bst->keys[ix] = p_bst->keys[ix + 1];
			p_bst->values[ix] = p_bst->values[ix + 1];
		}
		p_bst->size--;
		return p_bst->size;
	}
	else
		return -1;
}

void print(BinarySearchTable* p_bst) {
	int ix = 0;
	for (ix = 0; ix < p_bst->size; ix++) {
		printf("%c: %d\n", p_bst->keys[ix], p_bst->values[ix]);
	}
}

int main() {
	int key_rank;
	BinarySearchTable bst = { 0, 0, 0, 0 };
	initialize(&bst, 10);
	put(&bst, 'c', 50);
	put(&bst, 'a', 48);
	put(&bst, 'b', 49);
	key_rank = rank(&bst, 'd');
	print(&bst);
	printf("find: %d\n", key_rank);
	delete(&bst, 'a');
	print(&bst);
}

File: scenes/BinarySearchTable.py
from manim import *
from widgets.Struct import Struct

INTERVAL = 0.25
WIDTH = 0.5
SPACE = INTERVAL + WIDTH
UNIT_HEIGHT = 0.25
FONT_SIZE = 20
RUN_TIME = 0.5
ARROW_LENGTH = 0.5


class BinarySearchTable:
    def __init__(self, scene, capacity, posX=-config.frame_x_radius + INTERVAL, posY=-2):
        self.scene = scene
        self.capacity = capacity
        self.posX = posX
        self.posY = posY

        self.keys = []
        self.values = []
        self.objs = []

        self.rect = Rectangle(width=self.capacity * SPACE, height=3 * UNIT_HEIGHT + 0.2).move_to(
            [self.posX - INTERVAL / 2, self.posY, 0], LEFT)
        self.pointer = None
        self.lpointer = None
        self.hpointer = None

    def clearScene(self):
        animations = [Uncreate(self.rect, run_time=RUN_TIME), Uncreate(self.pointer, run_time=RUN_TIME)]
        for obj in self.objs:
            animations.append(Uncreate(obj, run_time=RUN_TIME))
        return animations

    def initScene(self, run_time=RUN_TIME):
        self.scene.play(Create(self.rect, run_time=run_time))

    def put(self, key, value):
        if len(self.keys) == 0 or key > self.keys[-1]:
            self.objs.append(Struct('BST', {'key': key, 'val': value}, unitWidth=WIDTH, unitHeight=UNIT_HEIGHT) \
                             .move_to([self.posX + len(self.keys) * SPACE, self.posY, 0], LEFT))
            self.keys.append(key)
            self.values.append(value)
            self.scene.play(Create(self.objs[-1], run_time=RUN_TIME))
        else:
            rank = self.rank(key)
            if rank < len(self.keys) and key == self.keys[rank]:
                oldobj = self.objs[rank]
                self.objs[rank] = Struct('BST', {'key': key, 'val': value}, unitWidth=WIDTH, unitHeight=UNIT_HEIGHT) \
                    .move_to(oldobj)
                self.keys[rank] = key
                self.values[rank] = value
                self.scene.play(*oldobj.setColor(RED))
                self.scene.wait(0.5)
                self.scene.play(ReplacementTransform(oldobj, self.objs[rank], run_time=RUN_TIME))
            else:
                obj = self.objs[rank]
                self.scene.play(*obj.setColor(RED))
                self.keys.append(key)
                self.values.append(value)
                self.objs.append(Struct('BST', {'key': key, 'val': value}, unitWidth=WIDTH, unitHeight=UNIT_HEIGHT))
                animations = []
                for i in range(len(self.keys) - 2, rank - 1, -1):
                    self.keys[i + 1] = self.keys[i]
                    self.values[i + 1] = self.values[i]
                    self.objs[i + 1] = self.objs[i]
                    animations.append(self.objs[i].animate(run_time=RUN_TIME).shift([SPACE, 0, 0]))
                if animations:
                    self.scene.play(*animations)
                self.keys[rank] = key
                self.values[rank] = value
                self.objs[rank] = Struct('BST', {'key': key, 'val': value}, unitWidth=WIDTH, unitHeight=UNIT_HEIGHT) \
                    .move_to([self.posX + SPACE * rank, self.posY, 0], LEFT)
                self.scene.play(Create(self.objs[rank], run_time=RUN_TIME), *obj.setColor(BLUE, RUN_TIME))

    def delete(self, key):
        if len(self.keys) == 0:
            return
        else:
            rank = self.rank(key)
            if rank < len(self.keys) and key == self.keys[rank]:
                oldobj = self.objs[rank]
                self.scene.play(*oldobj.setColor(RED))
                animations = []
                for i in range(rank, len(self.keys)-1):
                    self.keys[i] = self.keys[i+1]
                    self.values[i] = self.values[i+1]
                    self.objs[i] = self.objs[i+1]
                    animations.append(self.objs[i+1].animate(run_time=RUN_TIME).shift([-SPACE, 0, 0]))
                self.scene.play(FadeOut(oldobj, run_time=RUN_TIME))
                if animations:
                    self.scene.play(*animations)
                self.keys.pop(-1)
                self.objs.pop(-1)
                self.values.pop(-1)
            else:
                return

    def rank(self, key):
        if len(self.keys) == 0:
            return 0
        low = 0
        high = len(self.keys) - 1
        self.scene.wait(0.5)
        self.lpointer = self.generateLRPointer(LEFT, 'L').move_to(
            [self.posX + low * SPACE + WIDTH / 2, self.objs[0].get_critical_point(DOWN)[1] - SMALL_BUFF, 0], UP + RIGHT)
        self.rpointer = self.generateLRPointer(RIGHT, 'H').move_to(
            [self.posX + high * SPACE + WIDTH / 2, self.objs[0].get_critical_point(DOWN)[1] - SMALL_BUFF, 0], UP + LEFT)
        self.scene.play(FadeIn(self.lpointer, run_time=RUN_TIME), FadeIn(self.rpointer, run_time=RUN_TIME))
        while low <= high:
            mid = (low + high) // 2
            if self.pointer == None:
                self.pointer = self.generatePointer().move_to(
                    self.objs[mid].get_critical_point(UP) + [0, SMALL_BUFF, 0], DOWN)
                self.scene.play(Create(self.pointer, run_time=RUN_TIME))
            else:
                self.scene.play(self.pointer.animate(run_time=RUN_TIME).move_to(
                    self.objs[mid].get_critical_point(UP) + [0, SMALL_BUFF, 0], DOWN))
            if self.keys[mid] == key:
                self.scene.play(FadeOut(self.lpointer, run_time=RUN_TIME),
                                FadeOut(self.rpointer, run_time=RUN_TIME))
                self.lpointer = None
                self.rpointer = None
                return mid
            if self.keys[mid] > key:
                high = mid - 1
            else:
                low = mid + 1
            if low > high:
                self.lpointer[0].set_color(RED)
                self.rpointer[0].set_color(RED)
            self.scene.play(
                self.lpointer.animate(run_time=RUN_TIME).move_to(
                    [self.posX + low * SPACE + WIDTH / 2, self.objs[0].get_critical_point(DOWN)[1] - SMALL_BUFF, 0],
                    UP + RIGHT),
                self.rpointer.animate(run_time=RUN_TIME).move_to(
                    [self.posX + high * SPACE + WIDTH / 2, self.objs[0].get_critical_point(DOWN)[1] - SMALL_BUFF, 0],
                    UP + LEFT))
        self.scene.play(FadeOut(self.lpointer, run_time=RUN_TIME),
                        FadeOut(self.rpointer, run_time=RUN_TIME))
        self.lpointer = None
        self.rpointer = None
        return low

    def generatePointer(self):
        return Triangle(color=PINK, stroke_width=2).rotate(PI).scale(0.1)

    def generateLRPointer(self, direction, text='L', scale=0.2):
        if direction[0] == -1:
            pointer = VGroup()
            polygon = Polygon([0, 0, 0], [0, -2, 0], [1, -2, 0], [1, -1, 0],
                              fill_opacity=0, fill_color=WHITE, stroke_color=WHITE, stroke_width=2) \
                .flip() \
                .scale(scale)
            text = Text(text, font_size=18, fill_color=WHITE, fill_opacity=1) \
                .move_to(polygon.get_critical_point(DOWN), aligned_edge=DOWN)
            pointer.add(polygon, text)
            return pointer
        else:
            pointer = VGroup()
            polygon = Polygon([0, 0, 0], [0, -2, 0], [1, -2, 0], [1, -1, 0],
                              fill_opacity=0, fill_color=WHITE, stroke_color=WHITE, stroke_width=2) \
                .scale(scale)
            text = Text(text, font_size=18, fill_color=WHITE, fill_opacity=1) \
                .move_to(polygon.get_critical_point(DOWN), aligned_edge=DOWN)
            pointer.add(polygon, text)
            return pointer
